File: Robotics Kinematics and Dynamics Course/Chapter10/Lesson3/Chapter10_Lesson3.m
function dx = pendulum_rhs(t, x, params)
% x = [q; qd]
q  = x(1);
qd = x(2);

m = params.m;
l = params.l;
g = params.g;

% Example torque: PD about q = 0
Kp = 5.0;
Kd = 1.0;
tau = -Kp * q - Kd * qd;

qdd = (tau - m * g * l * sin(q)) / (m * l^2);

dx = [qd; qdd];
end

% Example of numerical integration:
params.m = 1.0;
params.l = 1.0;
params.g = 9.81;

x0 = [0.2; 0.0];  % initial angle and velocity
tspan = [0 10];

[t_sol, x_sol] = ode45(@(t, x) pendulum_rhs(t, x, params), tspan, x0);
      

File: Robotics Kinematics and Dynamics Course/Chapter10/Lesson3/Chapter10_Lesson3.cpp
#include <cmath>
#include <array>

// Simple pendulum dynamics: x = [q, qd]
struct PendulumParams {
  double m;
  double l;
  double g;
};

// User-specified torque profile
double torque(double t, double q, double qd) {
  // Example: simple PD control about q = 0
  double Kp = 5.0;
  double Kd = 1.0;
  return -Kp * q - Kd * qd;
}

// Right-hand side of the ODE: xdot = f(t, x)
std::array<double, 2> pendulum_rhs(double t,
                                    const std::array<double, 2>& x,
                                    const PendulumParams& p) {
  double q  = x[0];
  double qd = x[1];

  double tau = torque(t, q, qd);

  double denom = p.m * p.l * p.l;
  double qdd   = (tau - p.m * p.g * p.l * std::sin(q)) / denom;

  return {qd, qdd};
}
      
